staging/src/k8s.io/component-base/metrics/prometheus/health/metrics.go

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package health
+
+import (
+	"context"
+	"errors"
+
+	k8smetrics "k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+type HealthcheckStatus string
+
+const (
+	SUCCESS HealthcheckStatus = "success"
+	ERROR   HealthcheckStatus = "error"
+	PENDING HealthcheckStatus = "pending"
+)
+
+type HealthcheckType string
+
+const (
+	LIVEZ   HealthcheckType = "livez"
+	READYZ  HealthcheckType = "readyz"
+	HEALTHZ HealthcheckType = "healthz"
+)
+
+var (
+	// healthcheck is a Prometheus Gauge metrics used for recording the results of a k8s healthcheck.
+	healthcheck = k8smetrics.NewGaugeVec(
+		&k8smetrics.GaugeOpts{
```

Comment:
Using a gauge means we only get the most recent health check result, right?  If there are multiple health-checkers, will we potentially miss failures?

Did you consider using counters?

Reply:
> will we potentially miss failures?

It really depends on the scrape. But this is true for a counter with a restart as well. You should look at my `ObserveHealthcheck` method for the logic.

You can't use counters here since you don't know the polling frequency of the health endpoint (i.e. how often the kubelet is configured to hit it), since that's going to be where you end up updating the metric. And if someone manually hits the endpoint, that's going to screw up the delta math.

Reply:
Can you elaborate more on the math you plan to do?  I would've expected just to use a standard counter here (without resets), as that tells me the number of failed health-checks since the last scrape.

Reply:
You need to know how often the count is happening for the counter to be usable. Let's say you want to tolerate 1 failure per sixty seconds, but your healthcheck is being scraped every .5 seconds. In this case, you could have 2 failures over a scrape window but also be below your toleration threshold.

Reply:
> You need to know how often the count is happening for the counter to be usable. Let's say you want to tolerate 1 failure per sixty seconds, but your healthcheck is being scraped every .5 seconds. In this case, you could have 2 failures over a metric endpoint scrape window (every 60 seconds) but also be below your toleration threshold.

Reply:
But even so I added a counter just for you. I just don't it's actually usable.